ConcurrentAPI.adoc

```diff
@@ -0,0 +1,244 @@
+Parameters:
+time - the maximum time to wait for the lock
+unit - the time unit of the time argument
+Returns:
+true if the lock was acquired and false if the waiting time elapsed before the lock was acquired
+
+CompletableFuture fails with InfinispanLockException in case of error (InterruptedException, or any other non checked exceptions)
+
+###### CompletableFuture<Boolean> unlock();
+
+If the lock is rentrat, only the Thread holding the lock will be able to unlock, otherwise, any thread can unlock
+and it will behave as a Semaphore with one permit.
+true answer will say that the operation was succesul and the lock has been released, false the lock has not been relased
+
+
+=== Demo (as if the code was blocking)
+
+This main is an example on calling the API as if it was blocking, the corrent implementation should be done
+
+```java
+   public static void main(String[] args) throws Exception {
+        EmbeddedCacheManager cm = Infinispan.createClustered();
+        CounterManager counterManager = EmbeddedCounterManagerFactory.asCounterManager(cm);
+        DistributedLockManager lockManager = EmbeddedLockManagerFactory.asDistributedLockManager(cm);
+        InfinispanLock lock = lockManager.createLock("lock");
+        WeakCounter counter = counterManager.weakCounter("counter");
+
+        for (int i = 0; i < 100; i++) {
+            System.out.println("Counter on " + i + " is => " + counter.getValue());
+            lock.lock().get();
+            try {
+                counter.increment();
+                Thread.sleep(1000);
+                counter.decrement();
+            } finally {
+                lock.unLock();
+            }
+        }
+        cm.stop();
+    }
+
+```
+
+
+
+== (preview) Distributed CountDownLatch
+
+A synchronization aid that allows one or more threads to wait until a set of operations being performed in other threads completes.
+
+
+https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CountDownLatch.html
+
+
+=== CountDownLatchManager and configuration
+
+```java
+
+public class CountDownLatchManager {
+
+   InfinispanCountDownLatch define(CountDownLatchConfiguration configuration);
+   InfinispanCountDownLatch get(String name);
+   boolean remove(String name);
+
+}
+
+public interface InfinispanCountDownLatch {
+   CompletableFuture<Void> await();
+   CompletableFuture<Boolean> await(long timeout, TimeUnit unit);
+   CompletableFuture<Void> countDown();
+   CompletableFuture<Long> getCount();
+}
+
+
+public class CountDownLatchConfiguration {
+   private String name;
+   private int count;
+   private boolean createIfAbsent;
+   private boolean autoRemove;
+}
+
+ * @name A unique name
+ * @count initial count. If the countDownLatch already exists, override the current value
+ * @createIfAbsent if it does not exist, and the flag is set to true, then create it with the name and count
+ * @autoRemove when the countdown is zero, destroy it automatically.
+
+```
+
+=== Demo, just as if was locking interface
```

Comment:
Since you'd like to focus on clustered case, you could split this into two classed (Driver/Worker) with `main` method.